utilities/transactions/lock/range/range_locking_test.cc

```diff
@@ -0,0 +1,301 @@
+  // We verify that by checking the value of try_range_lock_called.
+  s = txn0->Put(cf, Slice("b"), Slice("value"), /*assume_tracked=*/true);
+  ASSERT_EQ(s, Status::OK());
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->ClearAllCallBacks();
+  ASSERT_EQ(try_range_lock_called, false);
+
+  txn0->Rollback();
+
+  delete txn0;
+}
+
+TEST_F(RangeLockingTest, SnapshotValidation) {
+  Status s;
+  Slice key_slice = Slice("k");
+  ColumnFamilyHandle* cfh = db->DefaultColumnFamily();
+
+  auto txn0 = NewTxn();
+  txn0->Put(key_slice, Slice("initial"));
+  txn0->Commit();
+
+  // txn1
+  auto txn1 = NewTxn();
+  txn1->SetSnapshot();
+  std::string val1;
+  s = txn1->Get(ReadOptions(), cfh, key_slice, &val1);
+  ASSERT_TRUE(s.ok());
+  val1 = val1 + std::string("-txn1");
+
+  s = txn1->Put(cfh, key_slice, Slice(val1));
+  ASSERT_TRUE(s.ok());
+
+  // txn2
+  auto txn2 = NewTxn();
+  txn2->SetSnapshot();
+  std::string val2;
+  // This will see the original value as nothing is committed
+  // This is also Get, so it is doesn't acquire any locks.
+  s = txn2->Get(ReadOptions(), cfh, key_slice, &val2);
+  ASSERT_TRUE(s.ok());
+
+  // txn1
+  s = txn1->Commit();
+  ASSERT_TRUE(s.ok());
+
+  // txn2
+  val2 = val2 + std::string("-txn2");
+  // Now, this call should do Snapshot Validation and fail:
+  s = txn2->Put(cfh, key_slice, Slice(val2));
+  ASSERT_TRUE(s.IsBusy());
+
+  s = txn2->Commit();
+  ASSERT_TRUE(s.ok());
+
+  /*
+    // Not meaningful if s.IsBusy() above is true:
+    // Examine the result
+    auto txn3= NewTxn();
+    std::string val3;
+    Status s3 = txn3->Get(ReadOptions(), cfh, key_slice, &val3);
+    fprintf(stderr, "Final: %s\n", val3.c_str());
+  */
+  delete txn0;
+  delete txn1;
+  delete txn2;
+}
+
+TEST_F(RangeLockingTest, MultipleTrxLockStatusData) {
+  WriteOptions write_options;
+  TransactionOptions txn_options;
+  auto cf = db->DefaultColumnFamily();
+  Status s;
+
+  Transaction* txn0 = db->BeginTransaction(write_options, txn_options);
+  Transaction* txn1 = db->BeginTransaction(write_options, txn_options);
+
+  // Get a range lock
+  s = txn0->GetRangeLock(cf, Endpoint("z"), Endpoint("z"));
+  ASSERT_EQ(s, Status::OK());
+
+  s = txn1->GetRangeLock(cf, Endpoint("b"), Endpoint("e"));
+  ASSERT_EQ(s, Status::OK());
+
+  auto s2 = range_lock_mgr->GetRangeLockStatusData();
+
+  ASSERT_EQ(s2.size(), 2);
+  delete txn0;
+  delete txn1;
+}
+
+TEST_F(RangeLockingTest, BasicLockEscalation) {
+  auto cf = db->DefaultColumnFamily();
+
+  auto counters = range_lock_mgr->GetStatus();
+
+  // Initially not using any lock memory
+  ASSERT_EQ(counters.current_lock_memory, 0);
+  ASSERT_EQ(counters.escalation_count, 0);
+
+  ASSERT_EQ(0, range_lock_mgr->SetMaxLockMemory(2000));
+
+  // Insert until we see lock escalations
+  auto txn0 = NewTxn();
+
+  // Get the locks until we hit an escalation
```

Comment:
It's not very clear to me when an escalation happens.
Does it happen when the lock memory threshold is exceeded?

Reply:
Is there an API to trigger lock escalation manually?

Reply:
Yes. It happens when the total amount of memory used for locks exceeds the limit specified by `SetMaxLockMemory`.
The memory is used for lock endpoints (whose size is known) + sizeof(certain_data_structures) so it is not possible to predict how many locks exactly one needs to take before they hit the limit.